### DISCUSSION TITLE
[c++] Open a mcap to append data

### DISCUSSION THREAD

**Discussion by ceccocats** [foxglove/mcap#1084] [General] [answered]
Is possible to open an existing mcap to write additional messages?
I can't find any interface to do this with McapWriter class in c++

**Comment by foxhubber** [comment]
Internal tracking ticket: [FG-2732](https://linear.app/foxglove/issue/FG-2732/[c]-open-a-mcap-to-append-data-[foxglovemcap-875])

**Comment by james-rms** [comment]
Not with the current client libraries, but such a thing is technically possible. Do you have a particular use case that requires this rather than opening a new MCAP file?

**Comment by ceccocats** [comment]
I need to write additional messages to mcap without rewrite everything.
Also open an mcap to rewrite could be useful in general.

**Reply by defunctzombie** [reply, marked as answer]
We tend to view mcap files as append-only, write once, robot log files rather than "database" files which you can go back and modify. If you want to write additional messages you can either use one of the reader libraries to read the existing file and write a new one, or use the CLI `mcap merge` command to merge several mcap files together.

**Reply by amacneil** [reply]
cc @wkalt who has investigated this request in the past

**Reply by ceccocats** [reply]
> We tend to view mcap files as append-only, write once, robot log files rather than "database" files which you can go back and modify. If you want to write additional messages you can either use one of the reader libraries to read the existing file and write a new one, or use the CLI `mcap merge` command to merge several mcap files together.

Following your philosophy it would be nice to have the possibility to append new data to an existing mcap, like you never closed the instance.

**Reply by gstorer** [reply]
I agree, the ability to append new data to existing file like it was never closed would be handy. This would allow a .mcap file to be treated the same way as a traditional text log file. The current API seems to be focused on running in sessions rather than long running processes that may be resumed.

**Reply by jtbandes** [reply]
FWIW, there is some prior art for doing this in Go in the `mcap` CLI tool: https://github.com/foxglove/mcap/pull/915

**Reply by jtbandes** [reply]
And related PR implementing append mode for TypeScript: https://github.com/foxglove/mcap/pull/1016

**Comment by defunctzombie** [comment]
> This would allow a .mcap file to be treated the same way as a traditional text log file. The current API seems to be focused on running in sessions rather than long running processes that may be resumed.

This is also by design to identify whether an mcap file was closed or that the recording process died before closing the file.

However as noted in another response, there are writers and approaches that can make this faster than re-writing the entire file - but not ones we've implemented in the current interfaces. If you are interested in append able mcap files I would suggest using un-chunked files (which are trivial to append to by over-writing the footer and adding more data) or using `mcap merge` on your files.
